hyperspy/learn/mva.py

```diff
@@ -435,97 +470,196 @@ def blind_source_separation(self,
             Any keyword arguments are passed to the BSS algorithm.
 
         """
-        target = self.learning_results
-        if not hasattr(target, 'factors') or target.factors is None:
-            raise AttributeError(
-                'A decomposition must be performed before blind '
-                'source seperation or factors must be provided.')
-        else:
-            if factors is None:
+        from hyperspy.signal import Signal
+        from hyperspy._signals.spectrum import Spectrum
+
+        lr = self.learning_results
+
+        # Check mask
+        if mask is not None:
+            ref_shape, space = (
+                (self.axes_manager.navigation_shape, "navigation")
+                if on_loadings
+                else
+                (self.axes_manager.signal_shape, "signal"))
+            if isinstance(mask, np.ndarray):
+                ref_shape = ref_shape[::-1]
+                if mask.shape != ref_shape:
+                    raise ValueError(
+                        "The `mask` shape is not equal to the %s shape in the"
+                        "array. Mask shape: %s\tSignal shape in array: %s" %
+                        (space, str(mask.shape), str(ref_shape)))
```

Comment:
I would change the last "Signal" to another `space` - it can be confusing with an error that states

"The `mask` shape is not equal to the __navigation__ shape in the array. Mask shape: (1,2,3)   __Signal__ shape in array: (3,2,1)"